R/growth_rate.R

```diff
@@ -195,71 +214,179 @@ growth_rate <- function(x = seq_along(y), y, x0 = x,
+      cli_abort(c(
+        "{.val {method}} requires all real values without missingness.",
+        i = "Set `na_rm = TRUE` and / or check for infinite values.",
+        i = "Using `log_scale = TRUE` may induce either case."
+      ))
+    }
 
-    # Smoothing spline
     if (method == "smooth_spline") {
-      params$x <- x
-      params$y <- y
-      obj <- do.call(stats::smooth.spline, params)
+      if (is.character(params$df)) params$df <- NULL
+      if (length(params$lambda) > 1L) {
+        cli_abort("{.val smooth_spline} requires 1 `lambda` but more were used.")
+      }
+      params <- params[c("df", "spar", "lambda", "cv", "all.knots", "df.offset", "penalty")]
+      params <- params[!sapply(params, is.null)]
+      obj <- rlang::inject(stats::smooth.spline(x = x, y = y, !!!params))
       f0 <- stats::predict(obj, x = x0)$y
       d0 <- stats::predict(obj, x = x0, deriv = 1)$y
       if (log_scale) {
         return(d0)
       } else {
         return(d0 / f0)
       }
-    } else {
-      # Trend filtering
-      ord <- params$ord
-      maxsteps <- params$maxsteps
-      cv <- params$cv
-      df <- params$df
-      k <- params$k
-
-      # Default parameters
-      ord <- ord %||% 3
-      maxsteps <- maxsteps %||% 1000
-      cv <- cv %||% TRUE
-      df <- df %||% "min"
-      k <- k %||% 3
-
-      # Check cv and df combo
-      if (is.numeric(df)) cv <- FALSE
-      if (!cv && !(is.numeric(df) && df == round(df))) {
-        cli_abort("If `cv = FALSE`, then `df` must be an integer.")
-      }
-
-      # Compute trend filtering path
-      obj <- genlasso::trendfilter(y = y, pos = x, ord = ord, max = maxsteps)
-
-      # Use CV to find df, if we need to
-      if (cv) {
-        cv_obj <- quiet(genlasso::cv.trendfilter(obj, k = k, mode = "df"))
-        df <- ifelse(df == "min", cv_obj$df.min, cv_obj$df.1se)
+    } else { # Trend filtering
+      params <- parse_trendfilter_params(params)
+      sdy <- stats::sd(y) # speeds computation significantly
+      y <- y / sdy
+      if (params$cv) {
+        obj <- trendfilter::cv_trendfilter(
+          y, x,
+          k = params$k, error_measure = params$error_measure,
+          nfolds = params$nfolds, family = params$family, lambda = params$lambda,
+          nlambda = params$nlambda, lambda_max = params$lambda_max,
+          lambda_min = params$lambda_min, lambda_min_ratio = params$lambda_min_ratio
+        )
+        lam <- params$df
+        which_lambda <- paste0("lambda_", lam)
+        f <- stats::predict(obj, newx = x0, which_lambda = which_lambda) * sdy
+      } else {
+        obj <- trendfilter::trendfilter(
+          y, x,
+          k = params$k, family = params$family, lambda = params$lambda,
+          nlambda = params$nlambda, lambda_max = params$lambda_max,
+          lambda_min = params$lambda_min, lambda_min_ratio = params$lambda_min_ratio
+        )
+        single_lambda <- length(obj$lambda == 1L)
```

Comment:
issue: bad parens

Reply:
I assume we need this logic because $dof isn't output for some reason if there is only one lambda?

Reply:
Good catch on the first. `$dof` is output in both cases. The question is "how did we get here". The user could have specified a sequence of lambda and a single `df` or a single lambda value (with `df = NULL`). In either case, we need to grab the single solution.